File: src/executor.py
from __future__ import annotations

import logging
from decimal import Decimal

from .adapters.base import VenueAdapter
from .models import Opportunity, Side, Venue

logger = logging.getLogger(__name__)


class Executor:
    def __init__(self, adapters: dict[str, VenueAdapter]) -> None:
        self.adapters = adapters

    async def execute_locked_arb(
        self,
        opp: Opportunity,
        venue_market_map: dict[tuple[str, str], str],
        size: Decimal,
        fragile_venue_first: bool = True,
    ) -> tuple[str | None, str | None]:
        yes_venue = opp.yes_venue.value
        no_venue = opp.no_venue.value
        yes_market = venue_market_map.get((yes_venue, opp.canonical_id), "")
        no_market = venue_market_map.get((no_venue, opp.canonical_id), "")

        if not yes_market or not no_market:
            logger.error("Missing market mapping for %s", opp.canonical_id)
            return None, None

        # Determine leg order: fragile (less liquid) first
        if fragile_venue_first and opp.yes_venue == Venue.KALSHI:
            first = ("yes", yes_venue, yes_market, opp.buy_yes_price, Side.YES)
            second = ("no", no_venue, no_market, opp.buy_no_price, Side.NO)
        elif fragile_venue_first and opp.no_venue == Venue.KALSHI:
            first = ("no", no_venue, no_market, opp.buy_no_price, Side.NO)
            second = ("yes", yes_venue, yes_market, opp.buy_yes_price, Side.YES)
        else:
            first = ("yes", yes_venue, yes_market, opp.buy_yes_price, Side.YES)
            second = ("no", no_venue, no_market, opp.buy_no_price, Side.NO)

        # Leg 1
        logger.info("Leg 1 (%s): %s %s at %s size %s",
                     first[0], first[1], first[2], first[3], size)
        try:
            first_order_id = await self.adapters[first[1]].place_order(
                first[2], first[4], first[3], size
            )
        except Exception as e:
            logger.error("Leg 1 failed: %s", e)
            return None, None

        if not first_order_id:
            logger.warning("Leg 1 returned no order ID — aborting")
            return None, None

        # Leg 2
        logger.info("Leg 2 (%s): %s %s at %s size %s",
                     second[0], second[1], second[2], second[3], size)
        try:
            second_order_id = await self.adapters[second[1]].place_order(
                second[2], second[4], second[3], size
            )
        except Exception as e:
            logger.error("Leg 2 failed after leg 1 filled: %s — RESIDUAL EXPOSURE", e)
            return first_order_id, None

        logger.info("Both legs executed: %s, %s | net_edge=%s",
                     first_order_id, second_order_id, opp.net_edge)
        return first_order_id, second_order_id

File: rust_arb/src/executor.rs
use anyhow::{Context, Result};
use rust_decimal::Decimal;
use tracing::{error, info, warn};

use crate::adapters::kalshi::KalshiAdapter;
use crate::adapters::polymarket::PolymarketAdapter;
use crate::models::{Opportunity, Side, Venue, VenueMapping};

pub struct Executor<'a> {
    pub kalshi: &'a KalshiAdapter,
    pub polymarket: &'a PolymarketAdapter,
}

impl<'a> Executor<'a> {
    pub fn new(kalshi: &'a KalshiAdapter, polymarket: &'a PolymarketAdapter) -> Self {
        Self { kalshi, polymarket }
    }

    pub async fn execute_locked_arb(
        &self,
        opp: &Opportunity,
        pm_mapping: &VenueMapping,
        km_mapping: &VenueMapping,
        size: Decimal,
    ) -> (Option<String>, Option<String>) {
        // Build legs: fragile venue (Kalshi) first
        let legs: Vec<(&str, Venue, &VenueMapping, Decimal, Side)> =
            if opp.yes_venue == Venue::Kalshi {
                vec![
                    ("yes", Venue::Kalshi, km_mapping, opp.buy_yes_price, Side::Yes),
                    ("no", Venue::Polymarket, pm_mapping, opp.buy_no_price, Side::No),
                ]
            } else if opp.no_venue == Venue::Kalshi {
                vec![
                    ("no", Venue::Kalshi, km_mapping, opp.buy_no_price, Side::No),
                    ("yes", Venue::Polymarket, pm_mapping, opp.buy_yes_price, Side::Yes),
                ]
            } else {
                vec![
                    ("yes", opp.yes_venue, pm_mapping, opp.buy_yes_price, Side::Yes),
                    ("no", opp.no_venue, pm_mapping, opp.buy_no_price, Side::No),
                ]
            };

        // Leg 1
        let (label1, venue1, mapping1, price1, side1) = &legs[0];
        info!("Leg 1 ({}): {} {} at {} size {}", label1, venue1, mapping1.native_market_id, price1, size);
        let first_id = match self.place_on_venue(*venue1, mapping1, *side1, *price1, size).await {
            Ok(id) if !id.is_empty() => id,
            Ok(_) => {
                warn!("Leg 1 returned no order ID — aborting");
                return (None, None);
            }
            Err(e) => {
                error!("Leg 1 failed: {}", e);
                return (None, None);
            }
        };

        // Leg 2
        let (label2, venue2, mapping2, price2, side2) = &legs[1];
        info!("Leg 2 ({}): {} {} at {} size {}", label2, venue2, mapping2.native_market_id, price2, size);
        let second_id = match self.place_on_venue(*venue2, mapping2, *side2, *price2, size).await {
            Ok(id) => id,
            Err(e) => {
                error!("Leg 2 failed after leg 1 filled: {} — RESIDUAL EXPOSURE", e);
                return (Some(first_id), None);
            }
        };

        info!(
            "Both legs executed: {}, {} | net_edge={}",
            first_id, second_id, opp.net_edge
        );
        (Some(first_id), Some(second_id))
    }

    async fn place_on_venue(
        &self,
        venue: Venue,
        mapping: &VenueMapping,
        side: Side,
        price: Decimal,
        size: Decimal,
    ) -> Result<String> {
        match venue {
            Venue::Kalshi => {
                self.kalshi
                    .place_order(&mapping.native_market_id, side, price, size)
                    .await
            }
            Venue::Polymarket => {
                let token_id = match side {
                    Side::Yes => mapping
                        .yes_token_id
                        .as_deref()
                        .context("missing yes_token_id for Polymarket")?,
                    Side::No => mapping
                        .no_token_id
                        .as_deref()
                        .context("missing no_token_id for Polymarket")?,
                };
                self.polymarket
                    .place_order(token_id, side, price, size, mapping.neg_risk)
                    .await
            }
        }
    }
}
